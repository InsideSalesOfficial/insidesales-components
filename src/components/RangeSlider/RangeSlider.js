import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import styled from 'styled-components';
import InputRange from 'react-input-range';
import './index.css';

import { colors, fontFamilies, fontSizes, fontWeights } from '../styles';

const RangeSliderWrap = styled.div`  
  font-family: ${fontFamilies.roboto};
  .input-range {
    .input-range__track {
      transition: left 0s, width 0s;
    }
    &:active {
      .input-range__track {
        transition: left 0.2s ease-out, width 0.2s ease-out;
      }
      .input-range__slider-container {
        transition: left 0.2s ease-out;
      }
    }
  }

  .input-range__slider {
    background-clip: padding-box;
    background-color: ${colors.green};
    
    border-radius: 50%;
    border-color: ${colors.green};

    box-sizing: border-box;
    margin-left: 0;
    margin-top: -7px;
    
    cursor: pointer;
    pointer-events: inherit;
    
    width: 12px;
    height: 12px;
    
    overflow: visible;
    outline: none;
  }

  .input-range__slider:hover {
    box-shadow: 0 0 0 5px ${colors.green20};
  }

  .input-range__track {
    background: ${colors.black12};
    height: 2px;
  }

  .input-range__track--active {
    background: ${colors.green};
  }

  .input-range__label {
    color: ${colors.black};
  }

  .input-range__label--min,
  .input-range__label--max {
    display: none;
  }

  .input-range__slider-container {
    transition: left 0s;
    &:active {
      transition: left 0.2s ease-out;
    }
  }

  .input-range__slider-container .input-range__label--value {
    left: 6px;
    top: -33px;
  }

  .input-range__slider-container:last-child .input-range__label--value {
    bottom: -31px;
    top: unset;
  }
`;

const Label = styled.label`
  color: ${colors.aluminum};
  display: block;
  font-size: ${fontSizes.xxSmall};
  font-weight: ${fontWeights.light};
  margin: 0 0 30px 0;
`;

/** A range slider UI element. */
class RangeSlider extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      value: 0
    };
  }

  componentWillMount() {
    if (this.props.initialValue) {
      this.setState({
        value: this.props.initialValue
      });
    }
  }

  componentWillReceiveProps(nextProps) {
    const { initialValue } = this.props;
    if (nextProps.initialValue && !_.isEqual(nextProps.initialValue, initialValue)) {
      this.setState({
        value: nextProps.initialValue
      });
    }
  }

  setValue = (value) => {
    this.setState({ value }, () => {
      this.props.onRangeUpdate(value);
    });
  }

  render() {
    return (
      <RangeSliderWrap className={this.props.className}>
        {this.props.label &&
          <Label>{this.props.label}</Label>
        }
        <InputRange
          formatLabel={this.props.formatLabel}
          maxValue={this.props.maxValue}
          minValue={this.props.minValue}
          onChange={this.setValue}
          className="pb-test__ranger-slider"
          step={this.props.step}
          value={this.state.value} />
      </RangeSliderWrap>
    );
  }
}

RangeSlider.defaultProps = {
  formatLabel: value => `${value}`,
  initialValue: 0,
  maxValue: 100,
  minValue: 0,
  onRangeUpdate: () => {},
  step: 1
};

RangeSlider.propTypes = {
  /** Function that is passed the slider handle's label and returns the label however user formats it */
  formatLabel: PropTypes.func,
  /** A single integer or an object of integers */
  initialValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.objectOf(PropTypes.number)
  ]),
  /** Function to execute when range value updates */
  onRangeUpdate: PropTypes.func,
  /** Label of above the range slider component */
  label: PropTypes.string,
  /** Maximum value of the range slide */
  maxValue: PropTypes.number,
  /** Minimum value of the range slider */
  minValue: PropTypes.number,
  /** Range step value for the step slider */
  step: PropTypes.number
};

export default RangeSlider;
