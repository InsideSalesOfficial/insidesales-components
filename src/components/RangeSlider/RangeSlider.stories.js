import React from "react";
import moment from "moment";
import styled from "styled-components";
import { action, storiesOf } from "@storybook/react";

import RangeSlider from "./RangeSlider";

import { colors, wrapComponentWithContainerAndTheme } from "../styles";
import { fontSizes } from "../styles";

const Hour = styled.span`
  font-size: ${fontSizes.small};
  margin-right: 4px;
`;

const Meridian = styled.small`
  font-size: ${fontSizes.xxSmall};
`;

function formatLabel(value) {
  return (
    <div>
      <Hour>
        {moment()
          .startOf("day")
          .add(value, "hours")
          .format("h")}
      </Hour>
      <Meridian>
        {moment()
          .startOf("day")
          .add(value, "hours")
          .format("A")}
      </Meridian>
    </div>
  );
}

function renderChapterWithTheme(theme) {
  return {
    info: `
        Usage

        ~~~
        import React from 'react';
        import {RangeSlider} from 'insidesales-components';
        ~~~
      `,
    chapters: [
      {
        sections: [
          {
            title: "Example: Single Range",
            subtitle: "A range slider with one value",
            sectionFn: () =>
              wrapComponentWithContainerAndTheme(
                theme,
                <RangeSlider
                  label="Pick a number"
                  minValue={0}
                  maxValue={100}
                  onRangeUpdate={action("rangeValues")}
                />,
                { height: "100px" }
              )
          },
          {
            title: "Example: Double Range",
            subtitle: "A range slider with a min and max value option",
            sectionFn: () =>
              wrapComponentWithContainerAndTheme(
                theme,
                <RangeSlider
                  label="Pick a number"
                  initialValue={{
                    min: 25,
                    max: 50
                  }}
                  minValue={0}
                  maxValue={100}
                  onRangeUpdate={action("rangeValues")}
                />,
                { height: "100px" }
              )
          },
          {
            title: "Example: Date Range",
            subtitle:
              "You can easily create a date range by formatting the labels.",
            info: `
                RangeSlider has a formatLabel prop which accepts a function where you can access the handle labels to format them

                ~~~
                const formatLabel = value => (
                  <div>
                    <Hour>{moment().startOf('day').add(value, 'hours').format('h')}</Hour>
                    <Meridian>{moment().startOf('day').add(value, 'hours').format('A')}</Meridian>
                  </div>
                );
                ~~~
              `,
            sectionFn: () =>
              wrapComponentWithContainerAndTheme(
                theme,
                <RangeSlider
                  label="time is between"
                  formatLabel={formatLabel}
                  initialValue={{
                    min: 8,
                    max: 17
                  }}
                  minValue={0}
                  maxValue={24}
                  onRangeUpdate={action("rangeValues")}
                />,
                { height: "100px" }
              )
          },
          {
            title: "Example: Custom Value Step",
            subtitle: "A range slider a step value of 10",
            sectionFn: () =>
              wrapComponentWithContainerAndTheme(
                theme,
                <RangeSlider
                  minValue={0}
                  maxValue={100}
                  onRangeUpdate={action("rangeValues")}
                  step={10}
                />,
                { height: "100px" }
              )
          }
        ]
      }
    ]
  };
}

storiesOf("Form", module)
  .addWithChapters("Default RangeSlider", renderChapterWithTheme({}))
  .addWithChapters(
    "RangeSlider w/ BlueYellow Theme",
    renderChapterWithTheme(colors.blueYellowTheme)
  );
