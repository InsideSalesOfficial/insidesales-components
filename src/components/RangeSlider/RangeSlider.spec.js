import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import RangeSlider from './RangeSlider';

describe('RangeSlider', () => {
  
  it('componentWillMount should set value on state if there is an initialValue prop', () => {
    const initialValue = 2;
    const wrapper = shallow(<RangeSlider minValue={0} maxValue={100} initialValue={initialValue} onRangeUpdate={jest.fn()} />);
    
    expect(wrapper.state().value).toEqual(initialValue);

  });

  it('componentWillReceiveProps should set value on state when the initialValue prop changes', () => {
    const initialValue = 2;
    const changedValue = 5;
    
    const wrapper = shallow(<RangeSlider minValue={0} maxValue={100} initialValue={initialValue} onRangeUpdate={jest.fn()} />);

    expect(wrapper.state().value).toEqual(initialValue);

    wrapper.setProps({initialValue: changedValue});

    expect(wrapper.state().value).toEqual(changedValue);
  });

  it('setValue set value on the state and call the onRangeUpdate prop after setState', () => {
    const onRangeUpdateMock = jest.fn();
    const wrapper = shallow(<RangeSlider minValue={0} maxValue={100} onRangeUpdate={onRangeUpdateMock} />);
    
    const state = wrapper.state();

    expect(state.value).toEqual(0);
    
    const newValue = 5;
    wrapper.instance().setValue(newValue);

    setTimeout(() => {
      expect(state.value).toEqual(newValue);
      expect(onRangeUpdateMock).toHaveBeenCalledWith(newValue);
    }, 0)
  });
});

describe('RangeSlider Snapshots', () => {
  test('shows component without a label', () => {
    const tree = renderer.create(<RangeSlider minValue={0} maxValue={100} onRangeUpdate={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  })
  test('shows component with a label', () => {
    const tree = renderer.create(<RangeSlider minValue={0} maxValue={100} label={'label'} onRangeUpdate={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  });
});
