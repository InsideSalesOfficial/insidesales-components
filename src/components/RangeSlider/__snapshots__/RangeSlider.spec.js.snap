// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider Snapshots shows component with a label 1`] = `
<div
  className="sc-bdVaJa llsCHA"
>
  <label
    className="sc-bwzfXH blRzDa"
  >
    label
  </label>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        0
      </span>
    </span>
    <div
      className="input-range__track input-range__track--background"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "0%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            0
          </span>
        </span>
        <div
          aria-controls={undefined}
          aria-labelledby={undefined}
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        100
      </span>
    </span>
  </div>
</div>
`;

exports[`RangeSlider Snapshots shows component without a label 1`] = `
<div
  className="sc-bdVaJa llsCHA"
>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        0
      </span>
    </span>
    <div
      className="input-range__track input-range__track--background"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "0%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            0
          </span>
        </span>
        <div
          aria-controls={undefined}
          aria-labelledby={undefined}
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        100
      </span>
    </span>
  </div>
</div>
`;
